contract_sale_invoicing/models/contract.py

```diff
@@ -9,23 +9,139 @@ class ContractContract(models.Model):
+             "contract")
+    group_by = fields.Selection([
+        ('sale_order', 'Sale Order'),
+        ('contract', 'Contract')],
+        default='sale_order', string='Create one invoice per')
 
-    @api.multi
-    def _recurring_create_invoice(self, date_ref=False):
-        invoices = super()._recurring_create_invoice(date_ref)
-        if not self.invoicing_sales:
-            return invoices
-        sales = self.env['sale.order'].search([
-            ('analytic_account_id', '=', self.group_id.id),
+    def get_sale_order_domain(self):
+        domain = [
             ('partner_invoice_id', 'child_of',
              self.partner_id.commercial_partner_id.ids),
             ('invoice_status', '=', 'to invoice'),
             ('date_order', '<=',
              '{} 23:59:59'.format(self.recurring_next_date)),
-        ])
-        if sales:
-            invoice_ids = sales.action_invoice_create()
-            invoices |= self.env['account.invoice'].browse(invoice_ids)[:1]
+        ]
+        if self.filter_with == 'analytic_account':
+            domain.append(('analytic_account_id', '=', self.group_id.id))
+        elif self.filter_with == 'contract':
+            domain.append(('contract_id', '=', self.id))
+        return domain
+
+    @api.multi
+    def _recurring_create_invoice(self, date_ref=False):
+        invoices = super()._recurring_create_invoice(date_ref)
+        for rec in self:
+            if not rec.invoicing_sales:
+                return invoices
+            so_domain = rec.get_sale_order_domain()
+            sales = self.env['sale.order'].search(so_domain)
+            if sales and rec.group_by == 'sale_order':
+                invoice_ids = sales.action_invoice_create()
+                invoices |= self.env['account.invoice'].browse(invoice_ids)[:1]
+
+    @api.multi
+    def _prepare_recurring_invoices_values(self, date_ref=False):
+        invoices_values = super()._prepare_recurring_invoices_values()
+        updated_invoices_values = []
+        for invoice_val in invoices_values:
+            invoice_line_list = []
+            for invoice_line in invoice_val.get('invoice_line_ids', []):
+                invoice_line = invoice_line[2] or {}
+                contract_line_rec = self.env['contract.line'].\
+                    browse(invoice_line.get('contract_line_id', False))
+                if contract_line_rec and contract_line_rec.contract_id and\
+                        contract_line_rec.contract_id.invoicing_sales and \
+                        contract_line_rec.contract_id.group_by == 'contract':
+                    so_domain = \
+                        contract_line_rec.contract_id.get_sale_order_domain()
+                    order_ids = self.env['sale.order'].search(
+                        so_domain, order='id asc')
+                    sale_order_line_product_qty = {}
+                    for order_id in order_ids:
+                        if not order_id.order_line.mapped('invoice_lines'):
```

Comment:
The existing of invoice lines doesn't mean that an order is totally invoiced. Keep in mind that an order can be partially invoiced.